### ISSUE TITLE
TMS34010 support

### ISSUE THREAD

**Issue by palmerj** [NationalSecurityAgency/ghidra#3990]
**Is your feature request related to a problem? Please describe.**

Please support for the tms34010 CPU. Is there a reason why this CPU hasn't been added yet? 

This CPU is used in many retro arcade games which users like to learn from or hack. 

Support info:

- See [Wiki](https://en.wikipedia.org/wiki/TMS34010)
- [User's guide](http://www.transputer.net/mtw/rg-750/doc/tms34010/t34010ug.pdf)
- [Instruction set](https://fabiensanglard.net/nbajamte/TMS34010_Instruction_Set.pdf)
- [Tech Datasheet](http://www.ti.com/lit/ds/symlink/tms34010.pdf)

**Describe the solution you'd like**
Add disassembly and processing analysis support for the tms34010 CPU

**Describe alternatives you've considered**
Other packages such as Radare2 or IDA pro do not support this either. There is http://www.mikekohn.net/micro/naken_asm.php, but best to have a nice IDE and GUI with all of the Ghidra features.

**Comment by ZPaul2Fresh8** [comment]
Currently working on it, new to sleigh so it's a learning curve for me.

**Comment by pjsoberoi** [comment]
Is there a disassembler available? If so I can try https://github.com/oberoisecurity/ghidra-processor-module-generator to autogenerate something.

**Comment by palmerj** [comment]
Thanks for the insightful comments @mumbel.

> No text (e/ oops not test) manual really sucks, are you aware of any that are not images (highlightable text)?

Maybe this might be more useful, but it's still not formatted that well https://archive.org/stream/bitsavers_tiTMS340xx988_22929190/TMS34010_Users_Guide_1988_djvu.txt ?

@pjsoberoi [naken](http://www.mikekohn.net/micro/naken_asm.php) can, but it's not supporting all instructions, and I'm not sure if it's dealing with bit addressing see https://github.com/mikeakohn/naken_asm/blob/master/docs/TMS340.md. Also [MAME](https://www.mamedev.org/) can dump the full rom disassembly via the debugger. The TMS arcade roms I'm interested in decompiling contain both data and opcode instructions are intermingled, so I'm not sure how well this might go. But it's worth a shot. But I'm also very interested in @ZPaul2Fresh8 work.

**Comment by mumbel** [comment]
@pjsoberoi   they're 16-bit instructions for the most part, but do have quite a few 16-bit or 2 16-bit values that follow (ie instructions can be 16, 32, or 48 bits) and then things like load/store multiple register instructions would be a lot of churn since its a 16-bit mask (but maybe just avoid all non 16-bit instructions in general).  I think if you could chop out some of those instructions and add after the fact it seems like a good candidate given a assembler hitting everything (I forget all the caveats the last time we talked about this).

**Comment by pjsoberoi** [comment]
@mumbel, 

Thanks for the suggestion. I did that (and lots of manual tweaks to GPMG). I have tested it, but a lot of instructions looks correct. A lot of instructions are doubled for the A register set vs the B register set. 

Instructions that look wrong:
- any instruction > 2 bytes (I ignored them)
- jr* (too many, didn't combine correctly)
- pixt* (gpmp couldn't handle the reg.XY syntax, didn't combine).

TL;DR: @ZPaul2Fresh8: My attached slaspec might save you some time. It has a lot of the grunt work of defining instructions completed.

[MyProc.slaspec.txt](https://github.com/NationalSecurityAgency/ghidra/files/8094583/MyProc.slaspec.txt)



.

**Comment by palmerj** [comment]
Interesting @pjsoberoi. What disassembly did you use?

**Comment by pjsoberoi** [comment]
I used naken_asm linked above. I wasted a ton of time trying to compile mame's implementation, couldn't get it to build.

**Comment by palmerj** [comment]
Thanks to @ZPaul2Fresh8 his WIP processor has been released https://github.com/ZPaul2Fresh8/Ghidra_TMS34010. Thanks Paul! I've reviewed the code, and it's pretty close to getting all of the opcode defined. A little more work needs to be done to define the MMTM and MMFM opcodes constructors, and most of the semantics still need to be done. 

In testing the processor, one thing annoying me is that the disassembly cannot display addresses correctly. As @mumbel noted above, the CPU uses bit addressing. Reading the [manual](https://fabiensanglard.net/nbajamte/t34010_user_guide.pdf) at section 3.1/2 I think the bit addresses range is from >0000 0000 to >FFFF FFFF (four gigabits) and is physically accessed 16 bits at a time. The bit address is the rightmost bit in the address word. For the programme memory addressing I'm looking at decompiling the base address is 0xFF800000, and range is 0xFF800000-0xFFFFFFFF. So to set the correct display addresses, the rom file bytes need to be loaded and mapped into the programme bit address at intervals of 0x10. 

Here is what the programme disassembly and memory addressing should look like:

<img width="433" alt="dis" src="https://user-images.githubusercontent.com/580022/168464622-0f889da5-d99f-462e-8514-750daf8382fe.png">

and here what the current disassembly in ghidra looks like with no offset or mapping to programme addresses:

<img width="898" alt="ghidra_dis" src="https://user-images.githubusercontent.com/580022/168464648-1bbbf812-8b04-4965-aff1-d71fa1fec1fe.png">

Of course, setting the offset address when loading the rom is easy to do via the GUI, but I can't see how to map the addresses address into the bit addresses range. Can any provide some guidance or pointers to example loader code to do this? I assume some java coding is going to be needed!

Many thanks

**Comment by ZPaul2Fresh8** [comment]
> Thanks to @ZPaul2Fresh8 his WIP processor has been released https://github.com/ZPaul2Fresh8/Ghidra_TMS34010. Thanks Paul! I've reviewed the code, and it's pretty close to getting all of the opcode defined. A little more work needs to be done to define the MMTM and MMFM opcodes constructors, and most of the semantics still need to be done.
> 
> In testing the processor, one thing annoying me is that the disassembly cannot display addresses correctly. As @mumbel noted above, the CPU uses bit addressing. Reading the [manual](https://fabiensanglard.net/nbajamte/t34010_user_guide.pdf) at section 3.1/2 I think the bit addresses range is from >0000 0000 to >FFFF FFFF (four gigabits) and is physically accessed 16 bits at a time. The bit address is the rightmost bit in the address word. For the programme memory addressing I'm looking at decompiling the base address is 0xFF800000, and range is 0xFF800000-0xFFFFFFFF. So to set the correct display addresses, the rom file bytes need to be loaded and mapped into the programme bit address at intervals of 0x10.
> 
> Here is what the programme disassembly and memory addressing should look like:
> 
> <img alt="dis" width="433" src="https://user-images.githubusercontent.com/580022/168464622-0f889da5-d99f-462e-8514-750daf8382fe.png">
> 
> and here what the current disassembly in ghidra looks like with no offset or mapping to programme addresses:
> 
> <img alt="ghidra_dis" width="898" src="https://user-images.githubusercontent.com/580022/168464648-1bbbf812-8b04-4965-aff1-d71fa1fec1fe.png">
> 
> Of course, setting the offset address when loading the rom is easy to do via the GUI, but I can't see how to map the addresses address into the bit addresses range. Can any provide some guidance or pointers to example loader code to do this? I assume some java coding is going to be needed!
> 
> Many thanks

Don't forget @mumbel , he has taught me a ton while working on this! I'm a noob with this stuff. Lately I've taken a liking to the hardware side of things and have been learning Verilog to recreate the logic of these classics!

**Comment by palmerj** [comment]
> Don't forget @mumbel , he has taught me a ton while working on this! I'm a noob with this stuff. 

Sorry I was unaware of @mumbel help. Massive thank you for your help to get it this far. And any further help here would be greatly appreciated :-)

> Lately I've taken a liking to the hardware side of things and have been learning Verilog to recreate the logic of these classics!

Yes!! Documenting that would be amazing, and really help with the hardware preservation of MK1-3 forever.

**Comment by mdentari** [comment]
Thanks for all the hard work on this. The TMS34010 has always fascinated me. It should have been a bigger hit in my opinion. Where can i send pizza money as a thanks?

**Comment by birdybro** [comment]
@ZPaul2Fresh8 I'm glad you are working on this. Many of us in the MiSTer FPGA community would like to see this chip reverse engineered for obvious reasons. If you need any help with the verilog implementation, our community has plenty of experienced FPGA devs willing to help.

**Comment by ZPaul2Fresh8** [comment]
> Any progress on this?

None, I've been focused on other tasks unfortunately. :(
